### PR TITLE
Migrate `emit-shared-files` and `emit-path-unhashed` `run-make` tests to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

try-job: x86_64-msvc
try-job: aarch64-apple
try-job: armhf-gnu
try-job: test-various

### CODE REVIEW

**Comment by jieyouxu** [comment]
I'll take a look later today, this looks like it warrants several try jobs.
@bors delegate+ (for try jobs only)

File: tests/run-make/emit-path-unhashed/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Specifying how rustc outputs a file can be done in different ways, such as
+// the output flag or the KIND=NAME syntax. However, some of these methods used
+// to result in different hashes on output files even though they yielded the
+// exact same result otherwise. This was fixed in #86045, and this test checks
+// that the hash is only modified when the output is made different, such as by
+// adding a new output type (in this test, metadata).
+// See https://github.com/rust-lang/rust/issues/86044
```
Comment (jieyouxu) [inline tests/run-make/emit-path-unhashed/rmake.rs:7]: Remark: very helpful comment 👍

File: tests/run-make/emit-shared-files/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    assert!(Path::new("invocation-only/settings.html").exists());
+    assert!(Path::new("invocation-only/x/all.html").exists());
+    assert!(Path::new("invocation-only/x/index.html").exists());
+    assert!(Path::new("invocation-only/theme-xxx.css").exists()); // generated from z.css
+    assert!(!Path::new("invocation-only/storage-xxx.js").exists());
+    assert!(!Path::new("invocation-only/SourceSerif4-It.ttf.woff2").exists());
+    // FIXME: this probably shouldn't have a suffix
+    assert!(Path::new("invocation-only/y-xxx.css").exists());
+    // FIXME: this is technically incorrect (see `write_shared`)
+    assert!(!Path::new("invocation-only/main-xxx.js").exists());
```
Comment (jieyouxu) [inline tests/run-make/emit-shared-files/rmake.rs:31]: Discussion: I know this is pre-existing, but what is happening here lol

**Review by jieyouxu — APPROVED** [review]
Thanks, the tests look very reasonable to me.

**Comment by jieyouxu** [comment]
r=me if the battery of try jobs come back green.

@bors try

File: tests/run-make/emit-shared-files/rmake.rs (hunk 2 above)
Comment (jieyouxu) [reply tests/run-make/emit-shared-files/rmake.rs:31]: cc @GuillaumeGomez sorry for the ping, but you reviewed the original PR #83478: do you happen to know what's the status for `--emit=toolchain-shared-resources` or what this test is trying to exercise here? In particular, I don't quite understand the FIXME even though I tried to read the original PR that introduced this test.

Comment (jieyouxu) [reply tests/run-make/emit-shared-files/rmake.rs:31]: Maybe <https://github.com/rust-lang/rust/issues/83784> is relevant?

File: src/tools/run-make-support/src/diff/mod.rs [hunk 3]
```diff
@@ -87,9 +87,7 @@ impl Diff {
         self
     }
 
-    #[track_caller]
-    pub fn run(&mut self) {
-        self.drop_bomb.defuse();
+    fn run_common(&self) -> (&str, &str, String, String) {
```
Comment (GuillaumeGomez) [inline src/tools/run-make-support/src/diff/mod.rs:90]: Would be better to return a struct with named field instead of a tuple here.

Comment (GuillaumeGomez) [reply src/tools/run-make-support/src/diff/mod.rs:90]: Hum although it's a private function, so not really a big issue either...

File: tests/run-make/emit-shared-files/rmake.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply tests/run-make/emit-shared-files/rmake.rs:31]: Ok I see. So normally, static files (like `main.js`) are generated with the hash of the file itself. I suppose the `--resource-suffix` option flag overrides this. Not sure if it's a good or bad though... Better to leave it as is.

File: src/tools/run-make-support/src/diff/mod.rs (hunk 3 above)
Comment (jieyouxu) [reply src/tools/run-make-support/src/diff/mod.rs:90]: Yeah I was also about to comment on that but I also realized that's private lol

**Pull request merged**